R/methods-epi_archive.R

```diff
@@ -360,6 +352,224 @@ epix_merge = function(x, y,
   ))
 }
 
+#' combine epi_archives by rows
+#'
+#' Take a sequence of archives and combine by rows. Complications arise if
+#' there are `time_value`s shared between the lists. `sync` determines how
+#' any later `NA`'s are treated, with the default `"forbid"` throwing an error,
+#' `"na"` treating them as intentional data (no modification), and `"locf"`
+#' filling forward across versions.
+#' Shared keys are another problem; by default, `force_distinct=FALSE`, meaning
+#' the entry in the earlier archive overwrites later archives. Otherwise there
+#' is an error on shared keys
+#' this function is still under active development, so there may be remaining
+#' edge cases
+#'
+#' @param ... list of `epi_archive` objects to append in order.
+#' @param sync Optional; `"forbid"`, `"na"`, or `"locf"`; in the case that later
+#'   versions contain `NA`'s, what do we do?
+#'   - `"forbid"`: emit an error if there are any shared time values between
+#'                     different archives;
+#'   - `"na"`: All `NA` values are treated as actual data, and thus are
+#'                 maintained (up to archival compression).
+#'   - `"locf"`: for every shared time value, use earlier versions of
+#'                   earlier archives to overwrite any `NA`'s found in later
+#'                   versions of later archives.
```

Comment:
I discussed the motivation somewhat in the main text, though it might make sense to move that conversation here. One way to mitigate the risk of overwriting meaningful `NA`s as well is to only `locf` on shared time values when the user specifically asks for it. Reflecting, there are possibly 3 levels to doing this writing:
1. overwrite all future `NA`'s, which is what `sync="locf"` currently does
2. only overwrite `NA`'s on shared `time_value`'s
3. only overwrite when the first entry in a given archive is `NA`; I think this is probably the best approach
2 and 3 are both somewhat more expensive, and definitely more complicated (you have to split the data, then perform the write on the problem half, and then rejoin).

Reply:
> sync="locf" is addressing a fairly subtle issue, where one indicator has versions released in a given month while the other doesn't. 
I think I'm starting to understand, but I think this situation probably would emerge only when a mistake has been made already in setting up the component archives, in `epix_merge` settings, or an `epix_merge` bug.  If signal A has version data where signal B does not, signal B should only have NA filled in if it's before its first measurement for a given epikey-time, or by filling forward explicit NAs for that epikey-time.  Is this the situation you're thinking of, and do you know if `epix_merge` has a bug where it's not doing this correctly?

Reply:
Oh, just saw [your great example](https://github.com/cmu-delphi/epiprocess/pull/343#issuecomment-1634942932).  I agree we need something like this functionality, maybe by default.

Some contributors to the problem:
- It's a bit unnatural for archives to hold partial version history, and that's part of the problem.
- The archive format not distinguishing explicit NAs vs. missing rows.
- Solving the point above is messier with our current `epix_merge()` implementation.  
- The other is not solving the explicit NA vs. missing row thing, and now it's even messier when things are merged.  At some point I thought I had an issue to use a lazy `epix_merge()` which would just store the parameters and re-implement some base archive operations; that might be useful if we want to make this cleaner later.